sfge/src/main/java/com/salesforce/rules/UnusedMethodRule.java

```diff
@@ -1,155 +1,159 @@
-            } else {
-                validator = new InstanceMethodCallValidator(candidateVertex, ruleStateTracker);
-            }
-            // If the validator can't find any usage, then we should add the method as unused.
-            if (!validator.usageDetected()) {
-                ruleStateTracker.trackUnusedMethod(candidateVertex);
+            String usageKey = generateUsageKey(methodVertex);
+
+            if (!encounteredUsageKeys.contains(usageKey)) {
+                String violationMsg =
+                        String.format(
+                                UserFacingMessages.RuleViolationTemplates.UNUSED_METHOD_RULE,
+                                methodVertex.getName(),
+                                methodVertex.getDefiningType());
+                results.add(
+                        new Violation.PathBasedRuleViolation(
+                                // NOTE: Since the violations represent the non-invocation of a
+                                // method, the method is both source and sink.
+                                violationMsg, methodVertex, methodVertex));
             }
         }
+        return results;
     }
 
-    /** Convert every known unused method to a violation, and return them in a list. */
-    private List<Violation> convertMethodsToViolations() {
-        return ruleStateTracker.getUnusedMethods().stream()
-                .map(
-                        m ->
-                                new Violation.StaticRuleViolation(
-                                        String.format(
-                                                VIOLATION_TEMPLATE,
-                                                m.getName(),
-                                                m.getDefiningType()),
-                                        m))
-                .collect(Collectors.toList());
+    /** Generate a unique key associated with this method, so we can mark its usage. */
+    private static String generateUsageKey(MethodVertex methodVertex) {
+        // Format of keys is "definingType#methodName@beginLine".
+        return methodVertex.getDefiningType()
+                + "#"
+                + methodVertex.getName()
+                + "@"
+                + methodVertex.getBeginLine();
     }
 
     /**
      * Returns true if the provided method isn't a valid candidate for analysis by this rule. Used
      * for filtering the list of all possible candidates into just the eligible ones.
      */
-    private boolean methodIsIneligible(MethodVertex vertex) {
-        // TODO: At this time, only static methods, constructors, and private instance methods are
-        //       supported. This limit will be loosened over time, and eventually removed entirely.
-        if (!vertex.isStatic() && !vertex.isConstructor() && !vertex.isPrivate()) {
-            return true;
-        }
-
+    @VisibleForTesting
+    public boolean methodIsIneligible(MethodVertex vertex) {
```

Comment:
Looks like some of these criteria could be added to the query so that they don't even show up in the results. The `directedToSkip` is probably the main case that cannot be filtered by the query.